File: multi_agent/projects/im-mobile/lib/services/reply_chain_service.dart
// Reply Chain Service for im-mobile

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/message_reply_chain.dart';

class ReplyChainService {
  static const String _baseUrl = 'http://localhost:8080/api/reply-chain';
  final String userId;
  final String nickname;

  ReplyChainService({required this.userId, required this.nickname});

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'X-User-Id': userId,
        'X-User-Nickname': nickname,
      };

  Future<MessageReplyChain> createReplyChain(ReplyChainRequest request) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/create'),
      headers: _headers,
      body: jsonEncode(request.toJson()),
    );
    if (response.statusCode == 200) {
      return MessageReplyChain.fromJson(jsonDecode(response.body));
    }
    throw Exception('Failed to create reply chain: ${response.body}');
  }

  Future<MessageReplyChain> getReplyChain(int chainId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/$chainId'),
      headers: _headers,
    );
    if (response.statusCode == 200) {
      return MessageReplyChain.fromJson(jsonDecode(response.body));
    }
    throw Exception('Failed to get reply chain');
  }

  Future<List<MessageReplyChain>> getConversationReplyChains(int conversationId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/conversation/$conversationId'),
      headers: _headers,
    );
    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((e) => MessageReplyChain.fromJson(e)).toList();
    }
    throw Exception('Failed to get conversation chains');
  }

  Future<MessageReplyChain> getBranchTree(int rootMessageId) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/branch/$rootMessageId'),
      headers: _headers,
    );
    if (response.statusCode == 200) {
      return MessageReplyChain.fromJson(jsonDecode(response.body));
    }
    throw Exception('Failed to get branch tree');
  }

  Future<void> markMessageDeleted(int messageId) async {
    final response = await http.put(
      Uri.parse('$_baseUrl/message/$messageId/deleted'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to mark message deleted');
    }
  }

  Future<void> deleteChain(int chainId) async {
    final response = await http.delete(
      Uri.parse('$_baseUrl/$chainId'),
      headers: _headers,
    );
    if (response.statusCode != 200) {
      throw Exception('Failed to delete chain');
    }
  }

  int calculateMaxDepth(MessageReplyChain chain) {
    if (chain.branchNodes.isEmpty) return chain.depth;
    int maxChild = 0;
    for (var node in chain.branchNodes) {
      final childDepth = calculateMaxDepthFromNode(node);
      if (childDepth > maxChild) maxChild = childDepth;
    }
    return maxChild > chain.depth ? maxChild : chain.depth;
  }

  int calculateMaxDepthFromNode(ReplyChainNode node) {
    return 0;
  }
}

File: im-desktop/src/services/reply-chain-service.ts
// Reply Chain Service for im-desktop

import { ReplyChain, ReplyChainRequest, BranchTree, MessageContext } from '../types/reply-chain';

const API_BASE = '/api/reply-chain';

class ReplyChainService {
  
  private getHeaders(): Record<string, string> {
    const userId = localStorage.getItem('userId') || '1';
    const nickname = localStorage.getItem('nickname') || 'Unknown';
    return {
      'Content-Type': 'application/json',
      'X-User-Id': userId,
      'X-User-Nickname': nickname,
    };
  }

  async createReplyChain(request: ReplyChainRequest): Promise<ReplyChain> {
    const response = await fetch(`${API_BASE}/create`, {
      method: 'POST',
      headers: this.getHeaders(),
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error('Failed to create reply chain');
    return response.json();
  }

  async getReplyChain(chainId: number): Promise<ReplyChain> {
    const response = await fetch(`${API_BASE}/${chainId}`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to get reply chain');
    return response.json();
  }

  async getConversationReplyChains(conversationId: number): Promise<ReplyChain[]> {
    const response = await fetch(`${API_BASE}/conversation/${conversationId}`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to get conversation reply chains');
    return response.json();
  }

  async getBranchTree(rootMessageId: number): Promise<ReplyChain> {
    const response = await fetch(`${API_BASE}/branch/${rootMessageId}`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to get branch tree');
    return response.json();
  }

  async getMessageContext(messageId: number): Promise<MessageContext> {
    const response = await fetch(`${API_BASE}/context/${messageId}`, {
      method: 'GET',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to get message context');
    return response.json();
  }

  async markMessageDeleted(messageId: number): Promise<void> {
    const response = await fetch(`${API_BASE}/message/${messageId}/deleted`, {
      method: 'PUT',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to mark message deleted');
  }

  async deleteChain(chainId: number): Promise<void> {
    const response = await fetch(`${API_BASE}/${chainId}`, {
      method: 'DELETE',
      headers: this.getHeaders(),
    });
    if (!response.ok) throw new Error('Failed to delete chain');
  }

  calculateMaxDepth(chain: ReplyChain): number {
    if (!chain.branchNodes || chain.branchNodes.length === 0) return chain.depth;
    return Math.max(chain.depth, ...chain.branchNodes.map(n => this.calculateMaxDepthFromNode(n)));
  }

  private calculateMaxDepthFromNode(node: ReplyChain): number {
    return node.depth || 0;
  }

  buildBranchPath(parentPath: string, messageId: number): string {
    return `${parentPath}/${messageId}`;
  }

  parseBranchPath(path: string): number[] {
    return path.split('/').filter(p => p && !isNaN(Number(p))).map(Number);
  }
}

export const replyChainService = new ReplyChainService();
